File: programmers/이중 우선순위 큐/Programmers42628_3.java
package com.example;

import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

public class Programmers42628_3 {
    // 제거하기 전에 최대힙 혹은 최소힙의 peek 의 값이 map 에 존재하는지 확인한다
    // 만약에 map 에 해당 key 값이 0이면 다른 힙에서 이미 제거된 상태다
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> minPQ = new PriorityQueue<>();
        PriorityQueue<Integer> maxPQ = new PriorityQueue<>(Comparator.reverseOrder());
        Map<Integer, Integer> map = new HashMap<>();
        for (String o : operations) {
            String[] operation = o.split(" ");
            String action = operation[0];
            int num = Integer.parseInt(operation[1]);
            if (action.equals("I")) {
                minPQ.add(num);
                maxPQ.add(num);
                map.put(num, map.getOrDefault(num, 0) + 1);
            } else {
                if (num == 1) {
                    removeInvalidNumFromPQ(maxPQ, map);
                    removeAndMinusNum(maxPQ, map);
                } else {
                    removeInvalidNumFromPQ(minPQ, map);
                    removeAndMinusNum(minPQ, map);
                }
            }
        }
        removeInvalidNumFromPQ(maxPQ, map);
        removeInvalidNumFromPQ(minPQ, map);
        return minPQ.isEmpty() || maxPQ.isEmpty()
                ? new int[]{0, 0}
                : new int[]{maxPQ.peek(), minPQ.peek()};
    }

    // 다른 PriorityQueue 에서 제거된 숫자를 해당 PriorityQueue 에서도 제거
    private void removeInvalidNumFromPQ(PriorityQueue<Integer> pq, Map<Integer, Integer> map) {
        while (!pq.isEmpty() && map.containsKey(pq.peek()) && map.get(pq.peek()) == 0) {
            pq.poll();
        }
    }

    // 유효한 숫자인 경우 PriorityQueue 에서 제거하고 Map 에서 값을 1 뺀다
    private void removeAndMinusNum(PriorityQueue<Integer> pq, Map<Integer, Integer> map) {
        if (!pq.isEmpty() && map.containsKey(pq.peek()) && map.get(pq.peek()) > 0) {
            Integer maxNum = pq.poll();
            map.put(maxNum, map.get(maxNum) - 1);
        }
    }

//    아래의 코드를 위의 코드로 리팩토링 했다
//    public int[] solution(String[] operations) {
//        PriorityQueue<Integer> minPQ = new PriorityQueue<>();
//        PriorityQueue<Integer> maxPQ = new PriorityQueue<>(Comparator.reverseOrder());
//        Map<Integer, Integer> map = new HashMap<>();
//        for (String o : operations) {
//            String[] operation = o.split(" ");
//            String action = operation[0];
//            int num = Integer.parseInt(operation[1]);
//            if (action.equals("I")) {
//                minPQ.add(num);
//                maxPQ.add(num);
//                map.put(num, map.getOrDefault(num, 0) + 1);
//            } else {
//                if (num == -1) {
//                    while (!minPQ.isEmpty() && map.containsKey(minPQ.peek()) && map.get(minPQ.peek()) == 0) {
//                        minPQ.poll();
//                    }
//                    if (!minPQ.isEmpty() && map.containsKey(minPQ.peek()) && map.get(minPQ.peek()) > 0) {
//                        Integer minNum = minPQ.poll();
//                        map.put(minNum, map.get(minNum) - 1);
//                    }
//                } else {
//                    while (!maxPQ.isEmpty() && map.containsKey(maxPQ.peek()) && map.get(maxPQ.peek()) == 0) {
//                        maxPQ.poll();
//                    }
//                    if (!maxPQ.isEmpty() && map.containsKey(maxPQ.peek()) && map.get(maxPQ.peek()) > 0) {
//                        Integer maxNum = maxPQ.poll();
//                        map.put(maxNum, map.get(maxNum) - 1);
//                    }
//                }
//            }
//        }
//        while (!maxPQ.isEmpty() && map.containsKey(maxPQ.peek()) && map.get(maxPQ.peek()) == 0) {
//            maxPQ.poll();
//        }
//        while (!minPQ.isEmpty() && map.containsKey(minPQ.peek()) && map.get(minPQ.peek()) == 0) {
//            minPQ.poll();
//        }
//        if (minPQ.isEmpty() || maxPQ.isEmpty()) {
//            return new int[]{0, 0};
//        }
//        return new int[]{maxPQ.peek(), minPQ.peek()};
//    }
}

File: programmers/이중 우선순위 큐/Programmers42628_3.kt
package com.example

class Programmers42628_3 {
    fun solution(operations: Array<String>): IntArray {
        val heap = IntervalHeap()
        for (o in operations) {
            val operation = o.split(" ")
            val action = operation[0]
            if (action == "I") {
                heap.add(operation[1].toInt())
            } else {
                if (heap.isEmpty()) {
                    continue
                }
                if (operation[1] == "-1") {
                    heap.pollMin()
                } else {
                    heap.pollMax()
                }
            }
        }
        if (heap.isEmpty()) {
            return intArrayOf(0, 0)
        }
        return intArrayOf(heap.getMax(), heap.getMin())
    }

    private class IntervalHeap {
        private val heap = mutableListOf(mutableListOf(-1, -1))

        fun isEmpty(): Boolean {
            return heap.size == 1
        }

        fun add(num: Int) {
            if (shouldAddArray()) {
                heap.add(mutableListOf(num))
            } else {
                heap[heap.lastIndex].add(num)
                if (isMinBiggerThanMax(heap.lastIndex)) {
                    switchElement(heap.lastIndex, 0, heap.lastIndex, 1)
                }
            }
            shiftUp()
        }

        private fun shouldAddArray(): Boolean {
            return heap.size == 1 || heap.last().size == 2
        }

        private fun shiftUp() {
            var curIdx = heap.lastIndex
            while (curIdx > 1) {
                val parentIdx = curIdx / 2
                if (heap[parentIdx].first() > heap[curIdx].first()) {
                    switchElement(parentIdx, 0, curIdx, 0)
                    curIdx = parentIdx
                } else if (heap[parentIdx].last() < heap[curIdx].last()) {
                    switchElement(parentIdx, 1, curIdx, heap[curIdx].lastIndex)
                    curIdx = parentIdx
                } else {
                    break
                }
            }
        }

        fun pollMin(): Int {
            // 첫번째와 마지막 인덱스의 첫번째 요소를 교환하고 마지막 인덱스의 첫번째 요소를 제거
            // 첫번째 인덱스의 첫번째 요소 위치 조정
            switchElement(1, 0, heap.lastIndex, 0)
            val num = heap.last().removeFirst()
            if (heap.last().isEmpty()) {
                heap.removeLast()
            }
            shiftDownMin(1)
            return num
        }

        private fun shiftDownMin(idx: Int) {
            var parentIdx = idx
            val leftChildIdx = parentIdx * 2
            val rightChildIdx = parentIdx * 2 + 1
            if (heap.size > leftChildIdx && heap[parentIdx].first() > heap[leftChildIdx].first()) {
                parentIdx = leftChildIdx
            }
            if (heap.size > rightChildIdx && heap[parentIdx].first() > heap[rightChildIdx].first()) {
                parentIdx = rightChildIdx
            }
            if (idx != parentIdx) {
                switchElement(idx, 0, parentIdx, 0)
                shiftDownMin(parentIdx)
            }
        }

        fun pollMax(): Int {
            switchElement(1, heap[1].lastIndex, heap.lastIndex, heap.last().lastIndex)
            val num = heap.last().removeLast()
            if (heap.last().isEmpty()) {
                heap.removeLast()
            }
            shiftDownMax(1)
            return num
        }

        private fun shiftDownMax(idx: Int) {
            var parentIdx = idx
            val leftChildIdx = parentIdx * 2
            val rightChildIdx = parentIdx * 2 + 1
            if (heap.size > leftChildIdx && heap[parentIdx].last() < heap[leftChildIdx].last()) {
                parentIdx = leftChildIdx
            }
            if (heap.size > rightChildIdx && heap[parentIdx].last() < heap[rightChildIdx].last()) {
                parentIdx = rightChildIdx
            }
            if (idx != parentIdx) {
                switchElement(idx, 1, parentIdx, heap[parentIdx].lastIndex)
                shiftDownMax(parentIdx)
            }
        }

        private fun isMinBiggerThanMax(idx: Int): Boolean {
            return heap.size > idx && heap[idx].size == 2 && heap[idx][0] > heap[idx][1]
        }

        private fun switchElement(idx1: Int, elementIdx1: Int, idx2: Int, elementIdx2: Int) {
            val temp = heap[idx1][elementIdx1]
            heap[idx1][elementIdx1] = heap[idx2][elementIdx2]
            heap[idx2][elementIdx2] = temp
        }

        fun getMin(): Int {
            return heap[1][0]
        }

        fun getMax(): Int {
            return heap[1].last()
        }
    }
}
